src/v/archival/tests/archiver_operations_impl_gtest.cc

```diff
@@ -1649,4 +1649,346 @@ TEST_CORO(
     }
 }
 
+enum class metadata_anomaly_type {
+    none = 0,
+    // gap between two segments
+    gap,
+    // overlap between two segments
+    overlap,
+    // the segment doesn't match record stats
+    record_stats_mismatch
+};
+
+ss::future<> test_admit_uploads_full_cycle(
+  std::deque<upload_result> results,
+  std::deque<metadata_anomaly_type> anomalies,
+  bool inline_manifest) {
+    scoped_config cfg;
+    cfg.get("storage_read_buffer_size").set_value(expected_read_buffer_size);
+    cfg.get("cloud_storage_segment_size_target")
+      .set_value(std::make_optional<size_t>(expected_target_size));
+    cfg.get("cloud_storage_segment_size_min")
+      .set_value(std::make_optional<size_t>(expected_min_size));
+    cfg.get("cloud_storage_bucket")
+      .set_value(std::make_optional(expected_bucket));
+
+    auto remote = ss::make_shared<remote_mock>();
+    auto pm = ss::make_shared<partition_manager_mock>();
+    auto builder = ss::make_shared<upload_builder_mock>();
+    auto partition = ss::make_shared<partition_mock>();
+
+    // set expectations
+    partition->expect_manifest(expected_manifest);
+
+    // Generate segment metadata and segments record stats
+    // which are mutually consistent if the 'add_anomaly' parameter is set to
+    // false or inconsistent otherwise.
+    auto gen_random_segment_meta = [](
+                                     const cloud_storage::segment_meta& prev,
+                                     metadata_anomaly_type add_anomaly) {
+        cloud_storage::segment_meta m;
+        m.is_compacted = false;
+        m.base_offset = model::next_offset(prev.committed_offset);
+        // Empty segments are disallowed and not tested
+        auto num_records = random_generators::get_int(1, 100);
+        auto num_configs = random_generators::get_int(num_records);
+        m.committed_offset = m.base_offset + model::offset(num_records);
+        m.delta_offset = prev.delta_offset_end;
+        m.delta_offset_end = m.delta_offset + model::offset_delta(num_configs);
+        m.size_bytes = random_generators::get_int(1, 0x8000);
```

Comment:
nit: please use base/units.h to write this as a human-readable number of bytes. Writing as hex is adds unneeded cognitive burden to all future readers

Reply:
it is 'some number' used in test, nobody needs to understand is it a MiB or KiB